Version/Version.cpp

```diff
@@ -0,0 +1,96 @@
+#include <filesystem>
+#include <Windows.h>
+#include <iostream>
+
+std::filesystem::path get_exe_path()
+{
+    TCHAR buffer[MAX_PATH];
+    GetModuleFileNameW(NULL, buffer, _countof(buffer));
+    return std::filesystem::path(buffer).parent_path();
+}
+
+std::filesystem::path find_api_directory()
+{
+    return get_exe_path()/(TEXT(R"(ArkApi)"));
+}
+
+std::filesystem::path find_dll()
+{
+    return find_api_directory()/(TEXT(R"(AsaApi.dll)"));
+}
+
+std::filesystem::path get_sys_path()
+{
+	TCHAR sys_dir[MAX_PATH];
+	GetSystemDirectoryW(sys_dir, MAX_PATH);
+	return std::filesystem::path(sys_dir);
+}
+
+std::filesystem::path find_vdll()
+{
+	return get_sys_path()/(TEXT(R"(Version.dll)"));
+}
+
+HINSTANCE m_hinst_dll = nullptr;
+HINSTANCE m_hinst_dll2 = nullptr;
+extern "C" UINT_PTR mProcs[17]{ 0 };
+
+LPCSTR import_names[] = {
+	"GetFileVersionInfoA", "GetFileVersionInfoByHandle", "GetFileVersionInfoExA", "GetFileVersionInfoExW",
+	"GetFileVersionInfoSizeA", "GetFileVersionInfoSizeExA", "GetFileVersionInfoSizeExW", "GetFileVersionInfoSizeW",
+	"GetFileVersionInfoW", "VerFindFileA", "VerFindFileW", "VerInstallFileA", "VerInstallFileW", "VerLanguageNameA",
+	"VerLanguageNameW", "VerQueryValueA", "VerQueryValueW"
+};
+
+BOOL WINAPI DllMain(HINSTANCE hinst_dll, DWORD fdw_reason, LPVOID)
+{
+	if (fdw_reason == DLL_PROCESS_ATTACH)
+	{
+		DisableThreadLibraryCalls(hinst_dll);
+
+		std::wstring vdllName = find_vdll().native();
+		m_hinst_dll = LoadLibraryW(vdllName.c_str());
+		if (m_hinst_dll == nullptr)
+			return FALSE;
+
+		for (int i = 0; i < 17; i++)
+		{
+			mProcs[i] = reinterpret_cast<UINT_PTR>(GetProcAddress(m_hinst_dll, import_names[i]));
+		}
+
+		std::wstring dllName = find_dll().native();
+		m_hinst_dll2 = LoadLibraryW(dllName.c_str());
+		if (m_hinst_dll2 == nullptr)
+			return FALSE;
+
+		using InitFuncType = void (*)();
+		InitFuncType InitApi = reinterpret_cast<InitFuncType>(GetProcAddress(m_hinst_dll2, "InitApi"));
+		if (InitApi)
+			InitApi();
+	}
+	else if (fdw_reason == DLL_PROCESS_DETACH)
+	{
+		FreeLibrary(m_hinst_dll2);
```

Comment:
I don't think that we should worry about ejecting the libraries here, as that is going to very likely cause a crash.

We've loaded and registered callbacks in the API (which we've just unloaded!) and every part of the game that requires a call to `version.dll` will also die!

On the other hand, this code should never **ever** run if Windows is working correctly as you generally can not eject a loaded in-use module...